x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/amazonbedrock/AmazonBedrockInferenceClientCache.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.amazonbedrock;
+
+import com.amazonaws.http.IdleConnectionReaper;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.inference.services.amazonbedrock.AmazonBedrockModel;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Collections.emptyMap;
+
+public final class AmazonBedrockInferenceClientCache implements AmazonBedrockClientCache {
+
+    private final BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator;
+    private volatile Map<Integer, AmazonBedrockBaseClient> clientsCache = emptyMap();
+
+    public AmazonBedrockInferenceClientCache(BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator) {
+        this.creator = creator;
+    }
+
+    public AmazonBedrockBaseClient getOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        var returnClient = internalGetOrCreateClient(model, timeout);
+        flushExpiredClients();
+        return returnClient;
+    }
+
+    private AmazonBedrockBaseClient internalGetOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        final Integer modelHash = AmazonBedrockInferenceClient.getModelKeysAndRegionHashcode(model, timeout);
+        {
+            final AmazonBedrockBaseClient client = clientsCache.get(modelHash);
+            if (client != null && client.tryToIncreaseReference()) {
+                return client;
+            }
+        }
+
+        synchronized (this) {
+            final AmazonBedrockBaseClient existing = clientsCache.get(modelHash);
+            if (existing != null && existing.tryIncRef()) {
+                return existing;
+            }
+
+            final AmazonBedrockBaseClient builtClient = creator.apply(model, timeout);
+
+            builtClient.mustIncRef();
+            clientsCache = Maps.copyMapWithAddedEntry(clientsCache, builtClient.hashCode(), builtClient);
+
+            return builtClient;
+        }
+    }
+
+    private void flushExpiredClients() {
+        var currentTimestampMs = System.currentTimeMillis();
```

Comment:
Should this be inside the synchronized block? Just thinking out loud: If some other thread acquired the lock and for some reason takes more time, the current thread has to wait and `currentTimestampMs` could become stale. If reading the time happens inside the synchronization block it should be a bit safer/avoid staleness, right? If yes we can also synchronize on the method level, because everything resides inside the synchronized block.

Reply:
Good call. It won't hurt to have it here even if the scenario you point out never happens...